pkg/driver/driver_gcp.go

```diff
@@ -0,0 +1,242 @@
+			},
+		})
+	}
+	instance.Disks = disks
+
+	var metadataItems = []*compute.MetadataItems{
+		{
+			Key:   "user-data",
+			Value: &d.UserData,
+		},
+	}
+	for _, metadata := range d.GCPMachineClass.Spec.Metadata {
+		metadataItems = append(metadataItems, &compute.MetadataItems{
+			Key:   metadata.Key,
+			Value: metadata.Value,
+		})
+	}
+	instance.Metadata = &compute.Metadata{
+		Items: metadataItems,
+	}
+
+	var networkInterfaces = []*compute.NetworkInterface{}
+	for _, nic := range d.GCPMachineClass.Spec.NetworkInterfaces {
+		computeNIC := &compute.NetworkInterface{
+			AccessConfigs: []*compute.AccessConfig{{}},
+		}
+		if len(nic.Network) != 0 {
+			computeNIC.Network = fmt.Sprintf("projects/%s/global/networks/%s", project, nic.Network)
+		}
+		if len(nic.Subnetwork) != 0 {
+			computeNIC.Subnetwork = fmt.Sprintf("regions/%s/subnetworks/%s", d.GCPMachineClass.Spec.Region, nic.Subnetwork)
+		}
+		networkInterfaces = append(networkInterfaces, computeNIC)
+	}
+	instance.NetworkInterfaces = networkInterfaces
+
+	var serviceAccounts = []*compute.ServiceAccount{}
+	for _, sa := range d.GCPMachineClass.Spec.ServiceAccounts {
+		serviceAccounts = append(serviceAccounts, &compute.ServiceAccount{
+			Email:  sa.Email,
+			Scopes: sa.Scopes,
+		})
+	}
+	instance.ServiceAccounts = serviceAccounts
+
+	operation, err := computeService.Instances.Insert(project, zone, instance).Context(ctx).Do()
+	if err != nil {
+		return "Error", "Error", err
+	}
+
+	if err := waitUntilOperationCompleted(computeService, project, zone, operation.Name); err != nil {
+		return "Error", "Error", err
+	}
+
+	return d.encodeMachineId(project, zone, d.MachineName), d.MachineName, nil
+}
+
+// Delete
+func (d *GCPDriver) Delete() error {
+	ctx, computeService, err := d.createComputeService()
+	if err != nil {
+		glog.Error(err)
+		return err
+	}
+
+	project, zone, name, err := d.decodeMachineId(d.MachineId)
+	if err != nil {
+		glog.Error(err)
+		return err
+	}
+
+	operation, err := computeService.Instances.Delete(project, zone, name).Context(ctx).Do()
+	if err != nil {
+		if ae, ok := err.(*googleapi.Error); ok && ae.Code == http.StatusNotFound {
+			return nil
+		}
+		glog.Error(err)
+		return err
+	}
+
+	return waitUntilOperationCompleted(computeService, project, zone, operation.Name)
+}
+
+// GetExisting
+func (d *GCPDriver) GetExisting() (string, error) {
+	return d.MachineId, nil
+}
+
+func (d *GCPDriver) createComputeService() (context.Context, *compute.Service, error) {
+	ctx := context.Background()
+
+	jwt, err := google.JWTConfigFromJSON(d.CloudConfig.Data["serviceAccountJSON"], compute.CloudPlatformScope)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	oauthClient := oauth2.NewClient(ctx, jwt.TokenSource(ctx))
+	computeService, err := compute.New(oauthClient)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return ctx, computeService, nil
+}
+
+func waitUntilOperationCompleted(computeService *compute.Service, project, zone, operationName string) error {
```

Comment:
Do we need this? Our controller is okay with a non-blocking (async) create call to the cloudProvider. The confirmation of create happens when the node object registers with the cluster.

Reply:
Not sure, so you don't receive an answer from GCP Go SDK regarding success of your request, but only an operation id which you can use to continuously ask for the state of it.
For example, while testing I did not enter a valid service account email. I did not receive an error here: https://github.com/gardener/node-controller-manager/pull/13/files/389e77556ff495c7efa4a9dec0b647ca853e8f4c..321f9396a3083ea602fb4789b42950c9a73d51d1#diff-af75cdff1c0efd51940289aeedf2ef4eR135
Thus, in the end I did not know whether an instance is getting created or not (without asking for the state of operation).
Same applies to deletion, without asking the operation for its state the driver would never know whether a instance has been successfully terminated or not. When it reports success but the VM still exist, we end up with tons of orphans.

Reply:
Update: We discussed regarding this over the call. 
We found that, different cloud-provider sdks manages the call analogues to createMachine differently. For example AWS-sdk immediately responds with error, if configuration provided to the call is buggy, on the other hand gcp does not throw any error immediately. gcp-sdk rather expects the user to check the status field of output object, after createMachine call.

We concluded that, we will keep the implementation different based on cloud-provider, considering the fact that this difference would be baked inside the driver-specific code, which is acceptable.